From 222b80dbb7ece7bd298e06f612156592f60a8efe Mon Sep 17 00:00:00 2001
From: Rajat Soni <quic_rajson@quicinc.com>
Date: Fri, 14 Apr 2023 15:02:11 +0530
Subject: [PATCH] ath12k: Add support to identify dual mac

While debugging using cnss diag for dual mac device, we are using Data.msc
file to get the logs. Instead we should be using Data_dual.msc file for dual
mac devices to get the logs.

Add a check if the device is dual mac and send it through nl command to be
sure that correct .msc file is used.

Signed-off-by: Rajat Soni <quic_rajson@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/testmode.c   | 10 ++++++++++
 drivers/net/wireless/ath/ath12k/testmode_i.h |  1 +
 2 files changed, 11 insertions(+)

diff --git a/drivers/net/wireless/ath/ath12k/testmode.c b/drivers/net/wireless/ath/ath12k/testmode.c
index 2e16255..96e8d19 100644
--- a/drivers/net/wireless/ath/ath12k/testmode.c
+++ b/drivers/net/wireless/ath/ath12k/testmode.c
@@ -26,6 +26,7 @@ static const struct nla_policy ath12k_tm_policy[ATH12K_TM_ATTR_MAX + 1] = {
 	[ATH12K_TM_ATTR_FWLOG]		= { .type = NLA_BINARY,
 					    .len = 2048 },
 	[ATH12K_TM_ATTR_LINK_IDX]		= { .type = NLA_U8 },
+	[ATH12K_TM_ATTR_DUAL_MAC]               = { .type = NLA_U8 },
 };
 
 void ath12k_fwlog_write(struct ath12k_base *ab, u8 *data, int len)
@@ -66,6 +67,15 @@ void ath12k_fwlog_write(struct ath12k_base *ab, u8 *data, int len)
 		}
 	}
 
+	if (ab->num_radios == 2)
+		ret = nla_put_u8(nl_skb, ATH12K_TM_ATTR_DUAL_MAC, ab->num_radios);
+
+	if (ret) {
+		ath12k_warn(ab, "failed to put dual mac wmi event to nl: %d\n", ret);
+		kfree_skb(nl_skb);
+		return;
+	}
+
 	cfg80211_testmode_event(nl_skb, GFP_ATOMIC);
 }
 
diff --git a/drivers/net/wireless/ath/ath12k/testmode_i.h b/drivers/net/wireless/ath/ath12k/testmode_i.h
index 3c6f8d0..0064770 100644
--- a/drivers/net/wireless/ath/ath12k/testmode_i.h
+++ b/drivers/net/wireless/ath/ath12k/testmode_i.h
@@ -27,6 +27,7 @@ enum ath12k_tm_attr {
 	ATH12K_TM_ATTR_WMI_OP_VERSION		= 6,
 	ATH12K_TM_ATTR_FWLOG                    = 7,
 	ATH12K_TM_ATTR_LINK_IDX			= 8,
+	ATH12K_TM_ATTR_DUAL_MAC			= 9,
 
 	/* keep last */
 	__ATH12K_TM_ATTR_AFTER_LAST,
-- 
2.17.1

